### DISCUSSION TITLE
Doing a re-write of qapply - any suggestions - generic evaluator?

### DISCUSSION THREAD

**Discussion by Costor** [sympy/sympy#24159] [Ideas]
My background is math, computer science and quantum computing.  I've just dug into qapply of the quantum package stumbling over issues #24142, #24153 and #24158 (and more unpublished). I also provided fixes for these, and also did a re-write of qapply for my purposes along the way.

Three questions: 

1. I've also seen (e.g. in #6268) that there might be reasonable concern about changing the behaviour of qapply. So it is worth while to continue a re-write? What would be conditions for a re-write to be accepted?
2. I feel there is no counterpart to qapply in the linear algebra packages of SymPy that would work with e.g. MatrixSymbol, generic vectors etc. the symbolic way qapply can with the quantum objects. Is that correct? If yes - what about thinking to carry over qapply to these objects too - actually both packages do linear algebra?
3. Pitfalls you can think of?

### How did I get there ###

Some days ago I wanted to do some quick computation of some controlled gates, i.e. compute a product of four tensor product operators (including OuterProduct |1><1| in some TensorProducts) with SymPy. 
I couldn't get a result from qapply and have already filed issues #24142, #24153 and #24158.  

I also found that qapply needed improvement in handling OuterProducts, powers of operators and some more. When I tried to fix this in the existing code I didn't see how to do it there.

So I re-wrote qapply as an generic expression evaluator, with a generic core that only knows about Add and Mul and is driven by properties and methods of the expression elements. By construction it also solves #6268 as it just follows the expression tree and distributes all forms of factors over all forms of sums explicitly without ever calling .expand(). I also limited recursion to the depth of the expression tree to avoid that larger powers or numbers of factors hit Python's recursion limit.

That sufficed for my initial computation task at hand, and it runs all test cases in test_qapply (and many more, with the exception of an additional expand() of the result), but of course it is fresh unproven code with some features missing. Nevertheless I consider it a reasonable basis as I feel the structure is clearer.

**Comment by oscarbenjamin** [comment]
> 1. So it is worth while to continue a re-write? What would be conditions for a re-write to be accepted?

I don't know the quantum code well but if you've looked at it and think that `qapply` realistically needs a rewrite then I'm not going to argue with that. Ideally the rewrite should preserve existing behaviour in so far as is possible (or desirable).

Since the quantum module is unmaintained there are probably lots of things that should be improved like this but if youi've looked through that code in detail then you probably know more about it than any of the current maintainers of SymPy. If you're interested in maintaining the quantum module and these are the sorts of improvements that are needed then that's fine by me.

> 2\. I feel there is no counterpart to qapply in the linear algebra packages

What exactly would a linear algebra version of `qapply` do?

The approach of using `qapply` in the quantum module is unusual compared to other parts of SymPy. I'm not sure to what extent `qapply` is only really needed just because the core isn't extensible. In other words because the authors of the quantum module wanted to use generic `Add` and `Mul` rather than making e.g. `QuantumAdd` and `QuantumMul` it wasn't possible to change the behaviour to account for the quantum-specific evaluation rules. Then `qapply` potentially exists as a workaround for a core that just isn't sufficiently extensible to apply those rules automatically.

The approach in the matrix expressions module is to have separate classes like `MatAdd`, `MatMul` etc. I haven't decided what I think of whether this is actually better than the `qapply` approach. Ideally there would be much less automatic evaluation in SymPy in general and then there would be lots of functions like `qapply` for specific kinds of manipulation and evaluation. The matrix expressions rely on `doit` for this but in general there can more than one possible thing to "do" so `doit` is not a general approach.

**Reply by Costor** [reply]
> In other words because the authors of the quantum module wanted to use generic Add and Mul rather than making e.g. QuantumAdd and QuantumMul it wasn't possible to change the behaviour to account for the quantum-specific evaluation rules. Then qapply potentially exists as a workaround for a core that just isn't sufficiently extensible to apply those rules automatically.

I had more or less assumed this, but it's good to hear that design decision from you: So qapply is essentially the 'evaluation kernel' of quantum expressions (QExpr class). 
I might even guess why the quantum package was not derived from matrices packages:
The matrices packages incorporate a basis and coefficients centric view to Linear Algebra, while quantum with its abstract coefficient-less definition of Ket (=abstract vector), Bra (=transposed and conjugated vector, the dual), Operator etc. didn't fit into that. However what quantum does is pure Linear Algebra, that's why I thought about to use the 'evaluation kernel' for coefficient-less standard Linear Algebra too.

>  If you're interested in maintaining the quantum module and these are the sorts of improvements that are needed then that's fine by me.

Actually I won't have the capacity to maintain quantum. As long as I continue in quantum computing and using a CAS to support my work, I will publish and try to fix quirks I come across but that would probably never touch quantum mechanics. 

With respect to qapply: given the role of qapply as evaluation kernel of quantum, qapply should be both performant and transparent, so its easy to extend. Also there are some quirks in it, and documented issues (e.g. the expand behaviour) which I feel can easier be adressed in a new design. So I will check whether the performance of a re-write is at least competitive, and if yes, present a PR for discussion.

P.S.: I've just read somewhere that quantum is now maintained. Is that right, who is the maintainer?

**Reply by oscarbenjamin** [reply]
> I've just read somewhere that quantum is now maintained

I think you might be reading a typo that I wrote earlier. To be clear the quantum module is *unmaintained*.

> it's good to hear that design decision from you

To be clear I haven't seen that design decision written down anywhere. It's just a guess on my part.

> Actually I won't have the capacity to maintain quantum

Being a "maintainer" here is not really an official thing to be clear. The main thing that is needed is someone who can be invited to review a PR or comment on an issue. Right now if you were to do any significant changes in the quantum module it would be hard for anyone else to really judge if those changes are a good idea or if they are even mathematically correct.

**Reply by Costor** [reply]
Ok, that definition of a maintainer sounds much more achievable. 

I've done some rough speed tests of my re-write of qapply and found it to be more than competitive in speed. And in terms of features it would be an improvement. 

In a local project environment I would now ask heavy duty users of the package to test the re-write. Any chance we can get hold of heavy duty users? Does anyone have an idea how many people are actually using qapply (I'm just asking, I understand that the answer is probably no - but maybe there is an indication from the number of questions concerning that package?)

Given we have no chance of intensive field testing, is there a chance to go parallel: 

- Replace qapply by the re-write but 
- provide the old qapply as qapply-classic.py so that `from qapply-classic import *` will restore the classic version?

**Reply by oscarbenjamin** [reply]
> Any chance we can get hold of heavy duty users?

You could try asking on the mailing list:
https://groups.google.com/g/sympy

I expect that the situation is that we can't really get hold of heavy duty users or that maybe there aren't that many in existence anyway.

We could provide independent functions for old and new qapply but what would be the benefit? I expect that it's probably better just to change it if the changes are an improvement.

The only thing I can really offer as a way to get more information is the list of open issues and PRs for the quantum label (you can also click to see the closed ones from there):
https://github.com/sympy/sympy/labels/physics.quantum

**Reply by oscarbenjamin** [reply]
> Ok, that definition of a maintainer sounds much more achievable.

I guess let's just try it and see what happens. We can tag you when a PR arises and if you find time then that's great but otherwise there is no formal commitment.

**Reply by Costor** [reply]
Ok, I'll do some polishing and tests and PR it in some days.
